modules/packages/Sort.chpl

```diff
@@ -796,6 +796,111 @@ module BinaryInsertionSort {
   }
 }
 
+pragma "no doc" 
+module TimSort {
+  private use Sort;
+
+    /*
+    Sort the 1D array `Data` using a parallel timSort algorithm.
+
+    :arg Data: The array to be sorted
+    :type Data: [] `eltType`
+    :arg SIZE: use :proc: `insertionSort` on blocks of `SIZE`
+    :type SIZE: `integral`
+    :arg comparator: :ref:`Comparator <comparators>` record that defines how the
+       data is sorted.
+
+   */
+
+  proc timSort(Data: [?Dom] ?eltType, SIZE=16, comparator:?rec=defaultComparator) {
+    chpl_check_comparator(comparator, eltType);
+
+    if Dom.rank != 1 {
+      compilerError("timSort() requires 1-D array");
+    }
+
+    _TimSort(Data, Dom.alignedLow, Dom.alignedHigh, SIZE, comparator);
+  }
+
+  private proc _TimSort(Data: [?Dom], lo:int, hi:int, SIZE=16, comparator:?rec=defaultComparator) {
+    import Sort.InsertionSort;
+
+    /*Parallely apply insertionSort on each block of size `SIZE`
+     using forall loop*/
+
+    const stride = if Dom.stridable then abs(Dom.stride) else 1;
+    var size = (hi - lo) / stride + 1;
+    var chunks:int = (size + SIZE - 1) / SIZE;
+    
+    forall i in 0..#chunks {
+      InsertionSort.insertionSort(Data, comparator = comparator, lo + (i * SIZE) * stride, min(hi, lo + ((i + 1) * SIZE * stride) - stride));
+    }
+
+    /* apply merge operations on each block
+    *as the merges at a level are independent of each other
+    *they can be applied parallely 
+    */
+
+    var num_size = SIZE;
+    while(num_size < size) {
+      forall i in 0..(size - 1) by 2 * num_size {
+
+        var l = lo + i * stride;
+        var mid = lo + (i + num_size - 1) * stride;
+        var r = min(lo + (i + 2 * num_size - 1) * stride, hi);
+
+        _Merge(Data, l, mid, r, comparator=comparator);
+      }
+
+      num_size = num_size * 2;
+
+    }
+  }
+
+  private proc _Merge(Dst: [?Dom] ?eltType, lo:int, mid:int, hi:int, comparator:?rec=defaultComparator) {
```

Comment:
Is this any different from `private proc _Merge` in `MergeSort` module? If not, I think we should make the `MergeSort` one public (but add a `pragma "no doc" so it doesn't show up in public docs), and re-use that, similar to how we re-use insertion sort.

Reply:
I have made few changes from `private proc _Merge` in `MergeSort`. In MergeSort it alternates the storage of Segments in the original memory and the copied memory to merge, but for parallelism, it was required to create a copy inside the `private proc _Merge` and store the results back into the original memory

Reply:
OK, then let's add a doc-string above this line to describe the diff between this `_Merge` and `MergeSort`'s version.

Reply:
```
    The `private proc _Merge` is disfferent from the `private proc _Merge` in `MergeSort`.
    In MergeSort it alternates the storage of segments in the original memory and the copied memory. 
    In the below defined `private proc _Merge`, it creates a copy of the segments to be merged and 
    store the results back into the original memory
```
Is this ok?

Reply:
Looks good, see some suggested edits below:

```
   This TimSort._Merge() differs from MergeSort._Merge() in the following way:
   MergeSort._Merge() alternates the storage of segments in the original memory and the copied memory. 
   TimSort._Merge() creates a copy of the segments to be merged and 
   stores the results back into the original memory.
```

Reply:
Done